sda-commons-server-jackson/src/main/java/org/sda/commons/server/jackson/hal/HalLinkInvocationStateUtility.java

```diff
@@ -0,0 +1,200 @@
+package org.sda.commons.server.jackson.hal;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Parameter;
+import java.lang.reflect.Proxy;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An utility class to process and store the invocation of a method. It also creates and returns a
+ * proxy that uses the method invocation handler.
+ */
+public class HalLinkInvocationStateUtility {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HalLinkInvocationStateUtility.class);
+
+  private static final ThreadLocal<MethodInvocationState> THREAD_LOCAL_INVOCATION_STATE =
+      ThreadLocal.withInitial(MethodInvocationState::new);
+
+  // Method Invocation Handler to process and save the current state of the method invocation
+  private static final InvocationHandler METHOD_PROCESSING_INVOCATION_HANDLER =
+      (proxy, method, methodArguments) -> {
+        // Get  invocation state from current thread
+        final MethodInvocationState methodInvocationState = THREAD_LOCAL_INVOCATION_STATE.get();
+        final String methodName = method.getName();
+        methodInvocationState.setInvokedMethod(methodName);
+        LOG.debug("Last invoked method '{}' added to current invocation state", methodName);
+        // Process annotated query and path parameters
+        processParams(methodInvocationState, methodArguments, method.getParameters());
+        methodInvocationState.processed();
+        // Do nothing
+        return null;
+      };
+
+  private HalLinkInvocationStateUtility() {}
+
+  /**
+   * Creates and returns a proxy instance based on the passed type parameter with a invocation
+   * handler, which processes and saves the needed method invocation information in the current
+   * thread. Parameters in the afterwards called method of the proxy will be used to resolve the URI
+   * template of the corresponding method. It should be ensured that the parameters are annotated
+   * with {@linkplain PathParam} or with {@linkplain QueryParam}.
+   *
+   * @param <T> the type parameter based on the passed type. Should not be null {@literal null}.
+   * @param type the type on which the method should be invoked.
+   * @return the proxy instance
+   */
+  public static <T> T methodOn(Class<T> type) {
+    return createProxy(type);
+  }
+
+  @SuppressWarnings("unchecked")
+  private static <T> T createProxy(Class<T> type) {
+    THREAD_LOCAL_INVOCATION_STATE.get().setType(type);
+    LOG.debug("Class type: '{}' added to the current invocation state", type);
+    return (T)
+        Proxy.newProxyInstance(
```

Comment:
Is it suitable for us that this only works with interfaces?

Reply:
Good point. I think java reflection doesn't provide that. Do you have any experience with [CGLIB](https://github.com/cglib/cglib)?

Reply:
Not so much as they have their caveats as well.
Maybe it is fine for the first step to document the requirements for using this kind of HalLinlkProvider more clearly.
For convenience I would prefer, that the entrypoint for the static methods of this utility (`methodOn(…)`) is in the HalLinkProvider that the consumers know. It can just delegate to here. I think it would be easier to have the entrypoint and all documentation in a single class.
From my experience with Spring Boot, I remember that such link builders can be very convenient but also require clear examples in the documentation as the API is not really fluent but needs a second entrypoint.

Reply:
The use of CGLIB allows us to use interfaces not exclusively. Classes are now also supported.

Reply:
But should we introduce a new dependency for that? We still need to document some details, e.g. the class requires a no-arg constructor.

Reply:
I'm not sure. We can force using an interface when the HalLinkProvider is used. In my opinion using an `interface` for the JAX-RS annotations is the preferred way in general. As i know for the services of mr-kaplan, we are using `interface` for all endpoint definitions. On the other hand cglib comes with light 300kb and two dependencies (not transitivee) but gives us the flexibility to use classes as well. 

Eather way, i should add some more documentation